### PR TITLE
Create Login Screen - Part 1

### PR DESCRIPTION
**Pull request by CuongPhan090**
Part1: Adding Login Presentation-Domain-Data layer + DI

### CODE REVIEW

File: app/build.gradle.kts [hunk 1]
```diff
@@ -18,6 +22,15 @@ android {
...
         }
+
+        // Load ApiKey from local properties
+        val localProperties = Properties()
+        val localPropertiesFile = rootProject.file("apikeys.properties")
+        if (localPropertiesFile.exists()) {
+            localProperties.load(localPropertiesFile.inputStream())
+        }
+        val apiKey = localProperties.getProperty("apiKey") ?: ""
+        buildConfigField("String", "API_KEY", "\"$apiKey\"")
```
Comment (philipplackner) [inline app/build.gradle.kts]: You can get the API key in one line nowadays:
```
val apiKey = gradleLocalProperties(rootProject.projectDir, providers).getProperty("API_KEY")
```

File: app/build.gradle.kts [hunk 2]
```diff
@@ -66,4 +80,19 @@ dependencies {
     androidTestImplementation("androidx.compose.ui:ui-test-junit4")
     debugImplementation("androidx.compose.ui:ui-tooling")
     debugImplementation("androidx.compose.ui:ui-test-manifest")
+
+    // Dagger Hilt
+    implementation("com.google.dagger:hilt-android:2.49")
+    kapt("com.google.dagger:hilt-android-compiler:2.46")
```
Comment (philipplackner) [inline app/build.gradle.kts]: I recommend to use KSP (Kotlin Symbol Processing), which is a faster alternative to kapt. Both Dagger and Room support that.

I explain it here: https://www.youtube.com/watch?v=WVhgUGyUnhQ

File: app/build.gradle.kts [hunk 3]
```diff
@@ -66,4 +80,19 @@ dependencies {
...
+    implementation("androidx.hilt:hilt-navigation-compose:1.2.0")
+
+    // Moshi
+    implementation("com.squareup.moshi:moshi-kotlin:1.15.0")
+
+    // Retrofit
+    implementation("com.squareup.retrofit2:retrofit:2.9.0")
+
+    // Moshi + Retrofit
+    implementation("com.squareup.retrofit2:converter-moshi:2.4.0")
```
Comment (philipplackner) [inline app/build.gradle.kts]: Which IDE version did you create this project on? Nowadays, the default project should use version catalogs for dependency management, have you looked into that?

File: app/src/main/java/com/cp/tasky/auth/login/domain/UserLoginValidationUseCase.kt [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+        return isValidEmail(email) && isValidPassword(password)
+    }
+
+    fun isValidEmail(email: String): Boolean {
+        return Patterns.EMAIL_ADDRESS.matcher(email).matches()
+    }
+
+    fun isValidPassword(password: String): Boolean {
+        return password.any { it.isLetter() && it.isLowerCase() } &&
+                password.any { it.isLetter() && it.isUpperCase() } &&
```
Comment (philipplackner) [inline app/src/main/java/com/cp/tasky/auth/login/domain/UserLoginValidationUseCase.kt]: Are there any characters that are lower case, but not letters? 😄 Or could you shorten this to just check for lower case?

File: app/src/main/java/com/cp/tasky/auth/shared/data/repository/AuthRepositoryImpl.kt [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    private fun LoginApiResponse.toLoginResponse(): LoginResponse {
+        return LoginResponse(
+            accessToken = accessToken,
+            refreshToken = refreshToken,
+            fullName = fullName,
+            userId = userId,
+            accessTokenExpirationTimestamp = accessTokenExpirationTimestamp,
+            message = message
+        )
+    }
```
Comment (philipplackner) [inline app/src/main/java/com/cp/tasky/auth/shared/data/repository/AuthRepositoryImpl.kt]: We typically put mappers in a separate file since it's a separate responsibility than the one from the repository (it calls the mappers, but doesn't define how mapping works)

File: app/src/main/java/com/cp/tasky/auth/shared/data/repository/RemoteDataSource.kt [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+import com.cp.tasky.auth.shared.data.remote.AuthApi
+import com.cp.tasky.auth.login.data.remote.models.LoginApiBody
+import com.cp.tasky.auth.login.data.remote.models.LoginApiResponse
+import javax.inject.Inject
+
+interface RemoteDataSource {
+
+    suspend fun loginUser(loginApiBody: LoginApiBody): LoginApiResponse
+}
```
Comment (philipplackner) [inline app/src/main/java/com/cp/tasky/auth/shared/data/repository/RemoteDataSource.kt:11]: Is there any value you gain from having this abstraction? In my experience, you rarely need abstractions where the abstraction and implementation are in the same layer. An abstraction only makes sense if you have or plan to have at least two implementations of it. Which other implementation do you plan to have here?

File: app/src/main/java/com/cp/tasky/ui/sharecomponent/ClickableText.kt [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    onTextClick: () -> Unit
+) {
+    val nonClickableString by remember {
+        mutableStateOf(
+            text.substring(
+                0,
+                text.indexOf(clickableText)
+            )
+        )
+    }
```
Comment (philipplackner) [inline app/src/main/java/com/cp/tasky/ui/sharecomponent/ClickableText.kt]: If you make this property a `val`, you say that it won't change. Why would you need to make it a state then, if it can't change?

File: app/src/main/java/com/cp/tasky/ui/sharecomponent/ClickableText.kt [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    val clickableString by remember { mutableStateOf(clickableText) }
```
Comment (philipplackner) [inline app/src/main/java/com/cp/tasky/ui/sharecomponent/ClickableText.kt]: Same here, the clickable text is just a hardcoded string or one from string resources. There's no need to make it a state if it can't change.

File: app/src/main/java/com/cp/tasky/ui/sharecomponent/PasswordTextField.kt [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+import androidx.compose.ui.text.input.PasswordVisualTransformation
+import androidx.compose.ui.text.input.VisualTransformation
+import androidx.compose.ui.tooling.preview.Preview
+import com.cp.tasky.R
+import com.cp.tasky.ui.theme.MEDIUM_8_DP
+
+
+@Composable
+fun PasswordTextField(modifier: Modifier = Modifier) {
+    var password by remember { mutableStateOf("") }
```
Comment (philipplackner) [inline app/src/main/java/com/cp/tasky/ui/sharecomponent/PasswordTextField.kt:28]: I would always define the states in a ViewModel, since yours will be reset to an empty string after a screen rotation or any other configuration change. In a ViewModel that can't happen and additionally you strictly separate the UI logic from the logic that decides about how the state looks like.

**Review by philipplackner — APPROVED** [review]

File: app/src/main/java/com/cp/tasky/auth/shared/data/repository/RemoteDataSource.kt (hunk 6 above)
Comment (CuongPhan090) [reply app/src/main/java/com/cp/tasky/auth/shared/data/repository/RemoteDataSource.kt:11]: I am planning to have registerUser here. Also, I want to have the abstraction to ease the testing process.

File: app/src/main/java/com/cp/tasky/ui/sharecomponent/ClickableText.kt (hunk 7 above)
Comment (CuongPhan090) [reply app/src/main/java/com/cp/tasky/ui/sharecomponent/ClickableText.kt]: you are correct, this nonClickableString will be never changed. it should not be state at all. What do you recommend to change this. 

I am thinking if doing this way: `    val nonClickableString = text.substring(0, text.indexOf(clickableText))` , this will compile every single time when the compose is recomposed. Somehow I need that `remember`

File: app/src/main/java/com/cp/tasky/ui/sharecomponent/ClickableText.kt (hunk 8 above)
Comment (CuongPhan090) [reply app/src/main/java/com/cp/tasky/ui/sharecomponent/ClickableText.kt]: The reason I don't want to have it in the string resource or hardcoded string is to have to function be more reusable, not tightly couple to this string only

File: app/build.gradle.kts (hunk 3 above)
Comment (CuongPhan090) [reply app/build.gradle.kts]: I am using HedgeHog 2023.1.1, but yes I will update to Koala 2024.1.1, and move dependency version to version catalogs
